text/src/tail.rs

```diff
@@ -67,25 +67,116 @@ impl Args {
     }
 }
 
-/// Prints the last `n` lines from the given buffered reader.
+/// Prints the last `n` lines from the given file.
 ///
 /// # Arguments
-/// * `reader` - A buffered reader to read lines from.
+/// * `file_path` - Path to the file.
 /// * `n` - The number of lines to print from the end. Negative values indicate counting from the end.
+fn print_last_n_lines<R: Read + Seek + BufRead>(reader: &mut R, n: isize) -> Result<(), String> {
+    if n < 0 {
+        // Print last `n` lines
+        let n = n.unsigned_abs();
+        let mut file_size = reader.seek(SeekFrom::End(0)).map_err(|e| e.to_string())?;
+        let mut buffer: VecDeque<String> = VecDeque::with_capacity(n);
+
+        let mut line = String::new();
+        let mut chunk_size = 1024;
+        let mut line_count = 0;
+
+        while file_size > 0 {
+            if file_size < chunk_size {
+                chunk_size = file_size;
+            }
+            reader
+                .seek(SeekFrom::Current(-(chunk_size as i64)))
+                .map_err(|e| e.to_string())?;
+            let mut buf = vec![0; chunk_size as usize];
+            reader.read_exact(&mut buf).map_err(|e| e.to_string())?;
+
+            for &byte in buf.iter().rev() {
+                if byte == b'\n' {
+                    if !line.is_empty() {
+                        buffer.push_front(line.clone());
+                        line.clear();
+                        if buffer.len() == n {
+                            break;
+                        }
+                    }
+                    line_count += 1;
+                }
+                line.insert(0, byte as char);
+            }
 
-fn print_last_n_lines<R: BufRead>(reader: R, n: isize) -> Result<(), String> {
-    let lines: Vec<_> = reader.lines().map_while(Result::ok).collect();
+            file_size = file_size.saturating_sub(chunk_size);
+            reader
+                .seek(SeekFrom::Current(-(chunk_size as i64)))
+                .map_err(|e| e.to_string())?;
+
+            if buffer.len() == n {
+                break;
+            }
+        }
+
+        // Handle case where the beginning of the file is reached but not all lines are counted
+        if !line.is_empty() && buffer.len() < n {
+            buffer.push_front(line);
+        }
 
-    let mut start = if n < 0 {
-        (lines.len() as isize + n).max(0) as usize
+        for line in buffer {
+            println!("{}", line.trim_end());
+        }
     } else {
-        (n - 1).max(0) as usize
-    };
+        let mut n = n;
+        if n == 0 {
+            n = 1;
+        }
+        // Print lines starting from the `n`-th line to the end
+        let mut line_count = 0;
+        let mut buffer = Vec::new();
+
+        // Count total lines in the file
+        while reader
+            .read_until(b'\n', &mut buffer)
```

Comment:
It's possible to have a very large single line. Since we're still reading to a vec here - we could encounter insufficient memory here.